tests/test_formatter.py

```diff
@@ -49,48 +48,118 @@ def test_python_code_with_multi_indent_passes(self):
+            '    return r"\t@RID"\n'
+        )
+        formatter = setup_formatter(python_code)
+        assert formatter.get_formatted() == python_code
+
+    def test_python_code_inside_run_keyword(self):
+        snake_code = (
+            "rule a:\n"
+            "    run:\n"
+            "        def s(a):\n"
+            "            if a:\n"
+            '                return "Hello World"\n'
+        )
+        formatter = setup_formatter(snake_code)
+        assert formatter.get_formatted() == snake_code
+
+    def test_snakemake_code_inside_python_code(self):
+        # The rules inside python code get formatted
+        formatter = setup_formatter(
+            "if condition:\n"
+            f"{TAB * 1}rule a:\n"
+            f'{TAB * 2}input: "a", "b"\n'
+            "else:\n"
+            f"{TAB * 1}rule b:\n"
+            f'{TAB * 2}script: "c.py"'
+        )
+        expected = (
+            "if condition:\n"
+            f"{TAB * 1}rule a:\n"
+            f"{TAB * 2}input:\n"
+            f'{TAB * 3}"a",\n'
+            f'{TAB * 3}"b",\n'
+            "else:\n"
+            f"{TAB * 1}rule b:\n"
+            f"{TAB * 2}script:\n"
+            f'{TAB * 3}"c.py"\n'
+        )
+        assert formatter.get_formatted() == expected
+
 
 class TestSimpleParamFormatting:
     def test_singleParamKeyword_staysOnSameLine(self):
         """
         Keywords that expect a single parameter do not have newline + indent
         """
-        formatter = setup_formatter("configfile: \n" '\t"foo.yaml"')
+        formatter = setup_formatter("configfile: \n" '    "foo.yaml"')
 
         actual = formatter.get_formatted()
-        expected = 'configfile: "foo.yaml" \n'
+        expected = 'configfile: "foo.yaml"\n'
 
         assert actual == expected
 
-    def test_singleParamKeywordInRule_staysOnSameLine(self):
+    def test_shell_keyword_get_newlineIndented(self):
+        formatter = setup_formatter(
+            "rule a:\n"
+            '    shell: "for i in $(seq 1 5);"\n'
+            '        "do echo $i;"\n'
+            '        "done"'
+        )
+        expected = (
+            "rule a:\n"
+            "    shell:\n"
+            '        "for i in $(seq 1 5);"\n'
+            '        "do echo $i;"\n'
+            '        "done"\n'
+        )
+        assert formatter.get_formatted() == expected
+
+    def test_singleParamKeywordInRule_NewlineIndented(self):
         formatter = setup_formatter(
             "rule a: \n"
-            '\tinput: "a", "b",\n'
-            '\t\t          "c"\n'
-            '\twrapper: "mywrapper"'
+            '    input: "a", "b",\n'
+            '                  "c"\n'
+            '    wrapper: "mywrapper"'
         )
 
         actual = formatter.get_formatted()
         expected = (
             "rule a:\n"
-            "\tinput: \n"
-            '\t\t"a", \n'
-            '\t\t"b", \n'
-            '\t\t"c", \n'
-            '\twrapper: "mywrapper" \n'
+            "    input:\n"
```

Comment:
Sorry, I know it's a pain, but should use `TAB` here to be explicit what we expect